src/gui/guiButton.cpp

```diff
@@ -749,76 +744,99 @@ void GUIButton::setColor(video::SColor color)
-			core::clamp<u32>(Colors[i].getRed() * COLOR_HOVERED_MOD, 0, 255),
-			core::clamp<u32>(Colors[i].getGreen() * COLOR_HOVERED_MOD, 0, 255),
-			core::clamp<u32>(Colors[i].getBlue() * COLOR_HOVERED_MOD, 0, 255));
-		PressedColors[i] = irr::video::SColor(Colors[i].getAlpha(),
-			core::clamp<u32>(Colors[i].getRed() * COLOR_PRESSED_MOD, 0, 255),
-			core::clamp<u32>(Colors[i].getGreen() * COLOR_PRESSED_MOD, 0, 255),
-			core::clamp<u32>(Colors[i].getBlue() * COLOR_PRESSED_MOD, 0, 255));
 	}
 }
-void GUIButton::setHoveredColor(video::SColor color)
-{
-	float d = 0.65f;
-	for (size_t i = 0; i < 4; i++) {
-		video::SColor base = Environment->getSkin()->getColor((gui::EGUI_DEFAULT_COLOR)i);
-		HoveredColors[i] = base.getInterpolated(color, d);
-	}
-}
-void GUIButton::setPressedColor(video::SColor color)
+
+//! Set element properties from a StyleSpec corresponding to the button state
+void GUIButton::setFromState()
 {
-	float d = 0.65f;
-	for (size_t i = 0; i < 4; i++) {
-		video::SColor base = Environment->getSkin()->getColor((gui::EGUI_DEFAULT_COLOR)i);
-		PressedColors[i] = base.getInterpolated(color, d);
-	}
+	StyleSpec::State state = StyleSpec::STATE_DEFAULT;
+
+	if (isPressed())
+		state = static_cast<StyleSpec::State>(state | StyleSpec::STATE_PRESSED);
+
+	if (isHovered())
+		state = static_cast<StyleSpec::State>(state | StyleSpec::STATE_HOVERED);
+
+	setFromStyle(StyleSpec::getStyleFromStatePropagation(Styles, state));
 }
 
 //! Set element properties from a StyleSpec
-void GUIButton::setFromStyle(const StyleSpec& style, ISimpleTextureSource *tsrc)
+void GUIButton::setFromStyle(const StyleSpec& style)
 {
+	bool hovered = (style.getState() & StyleSpec::STATE_HOVERED) != 0;
+	bool pressed = (style.getState() & StyleSpec::STATE_PRESSED) != 0;
+
 	if (style.isNotDefault(StyleSpec::BGCOLOR)) {
+
 		setColor(style.getColor(StyleSpec::BGCOLOR));
-	}
-	if (style.isNotDefault(StyleSpec::BGCOLOR_HOVERED)) {
-		setHoveredColor(style.getColor(StyleSpec::BGCOLOR_HOVERED));
-	}
-	if (style.isNotDefault(StyleSpec::BGCOLOR_PRESSED)) {
-		setPressedColor(style.getColor(StyleSpec::BGCOLOR_PRESSED));
+
+		// If we have a propagated hover/press color, we need to automatically
+		// lighten/darken it
+		if (!Styles[style.getState()].isNotDefault(StyleSpec::BGCOLOR)) {
+			for (size_t i = 0; i < 4; i++) {
+				if (pressed) {
+					Colors[i] = irr::video::SColor(Colors[i].getAlpha(),
+							core::clamp<u32>(Colors[i].getRed() * COLOR_PRESSED_MOD, 0, 255),
+							core::clamp<u32>(Colors[i].getGreen() * COLOR_PRESSED_MOD, 0, 255),
+							core::clamp<u32>(Colors[i].getBlue() * COLOR_PRESSED_MOD, 0, 255));
+				} else if (hovered) {
+					Colors[i] = irr::video::SColor(Colors[i].getAlpha(),
+							core::clamp<u32>(Colors[i].getRed() * COLOR_HOVERED_MOD, 0, 255),
+							core::clamp<u32>(Colors[i].getGreen() * COLOR_HOVERED_MOD, 0, 255),
+							core::clamp<u32>(Colors[i].getBlue() * COLOR_HOVERED_MOD, 0, 255));

```

Comment:
This color modifying & clamping code should become a helper function.
I think either `numeric.cpp` or `StyleSpec.cpp` would be suited.

Reply:
Good call, that was starting to get too repetitive.
I don't like putting this in StyleSpec because it's not closely-related enough (this is more default element behavior than styling behavior), though putting the function in numeric adds an extra include.

Not sure how I feel about that, but it's better than what was there before so I've done it.